index.rst

```diff
@@ -13,17 +13,256 @@
+It would involve considerable code changes, but more removals than additions, and the vast majority of these would be behind the scenes or of minimal impact to users (e.g. ``Butler.datastore`` and ``Registry.insertDatasets`` are not formally private, but they should be, and are already widely recognized as for internal use only).
+
+Implementation of important butler operations
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In order for this proposal to be considered viable, we need to look carefully at how we would implement joint Registry/Datastore operations that we want to maintain consistency.
+Note that later sections in this technical note will expand upon these and ultimately take the form of prototype Python code, as further changes are added to the proposal.
+
+``Butler.put``
+""""""""""""""
+
+1. Obtain a valid expanded data ID from the Registry.
+   In the vast majority of cases (i.e. QuantumGraph execution) this done well in advance of the actual ``put`` call.
+
+2. Construct a ``DatasetRef`` by generating an appropriate UUID and using an existing or soon-to-exist RUN name.
+   This will also typically occur well before the rest of the ``put``, as part of QuantumGraph generation.
+
+3. Perform the ``Datastore.put`` operation, writing the file artifacts associated with the dataset and returning records to the Butler.
+   Datastore can be expected to make this operation atomic, either because it is naturally atomic for its storage backing or via writing a temporary and moving it.
+   We do have to accept the possibility of failures leaving partially-written temporary files around.
+   When the Datastore is backed by storage that requires signed URLs, this write is to a staging area and uses non-signed URIs.
+
+4. If the butler has a Registry, either held directly (as in a "full butler" today) or as a client of a butler server, call ``Registry.insertDatasets`` with both the ``DatasetRef`` and the records returned by the Datastore.
+   Database transactions can be used to ensure that all tables in the Registry (including those for datastore records) are updated consistently or left unchanged.
+   If this operation fails, or the butler does not have a Registry (e.g. ``QuantumBackedButler``), the dataset is left in a valid state: it is in the Datastore, but not the Registry.
+   If the Datastore is backed by storage that requires signed URLs, this call also tells the server to transfer the artifacts from the staging area to permanent storage after checking for permissions.
+   This must happen before the database changes are committed.
+
+This has two major advantages over our current ``put`` implementation:
+
+- there is no database transaction over the Datastore write, keeping transactions small and reducing contention for database connections;
```

Comment:
Is datastore still meant to clean up after itself? If it writes 4 files and somehow fails to write the 5th, should it attempt to delete the previous four?

Reply:
I don't think so, but that isn't a strong opinion. If we have some scenario or use case that makes that kind of behavior common, then it might be convenient to clean up immediately, and there's certainly nothing wrong with trying to clean up immediately.  But we can't rely on it to guarantee consistency, and my _suspicion_ is that once we have ways to clean up orphaned Datastore-only files, the automatic-undo logic won't be pulling its weight in terms of code complexity and maintenance.

Reply:
We might have to reorganize datastore writes so that it writes the N temporary files for disassembled composites and then renames (or uploads) them at the end.